predicators/envs/spot_env.py

```diff
@@ -671,7 +752,48 @@ def _get_next_nonpercept_atoms(self, obs: _SpotObservation,
         }
 
     def simulate(self, state: State, action: Action) -> State:
-        raise NotImplementedError("Simulate not implemented for SpotEnv.")
+        assert isinstance(action.extra_info, (list, tuple))
+        action_name, action_objs, _, _, action_fn, action_fn_args = \
+            action.extra_info
+        # The extra info is (action name, objects, function, function args).
+        # The action name is either an operator name (for use with nonpercept
+        # predicates) or a special name. See below for the special names.
+        assert self.action_space.contains(action.arr)
+
+        # Update the poses of the robot and all relevant objects
+        # to be in the correct position(s).
+        update_pbrspot_given_state(self.sim_robot, _obj_name_to_sim_obj, state)
+
+        # Special case: the action is "done", indicating that the robot
+        # believes it has finished the task. Used for goal checking.
+        if action_name == "done":
+            # During planning, trust that the goal is accomplished if the
+            # done action is returned.
+            self._current_task_goal_reached = True
+            return state
+
+        # Execute the action in the pybullet env. Automatically retry
+        # if a retryable error is encountered.
+        action_fn(*action_fn_args)  # type: ignore
+
+        # Construct a new state given the updated simulation.
+        next_state = construct_state_given_pbrspot(self.sim_robot,
+                                                   _obj_name_to_sim_obj, state)
+
+        # In the future, we'll probably actually implement proper ways to
+        # check things like InHandView and other quantities below
+        # in the simulator (e.g., by shooting a ray!). However for now,
+        # we're going to just hack things such that the expected atoms check
+        # passes. This will be done in a manner similar to dry sim.
+        if action_name == "MoveToHandViewObject":
+            obj_to_view = action_objs[1]
+            next_state.set(obj_to_view, "in_hand_view", 1.0)
+        elif action_name == "PickObjectFromTop":
```

Comment:
are these two just the ones that have been implemented so far? should we add an assertion that no other action is being used yet?